### ISSUE TITLE
Proposal: add config blob discovery

### ISSUE THREAD

**Issue by kyessenov** [envoyproxy/envoy#11547]
This proposal is to add a generic Blob Discovery Service that allows the control plane to push arbitrary data blobs and keep them up to date without an update churn using resource versions. The motivation is Wasm extension files which are realistically on the order of MBs. The current solutions have many issues:
- inlining MB sized binaries in LDS/RDS config is problematic and can cause head-of-line blocking in ADS;
- URL fetcher has no interaction with xDS subscriptions. That means first xDS update is always NACKed. URL fetches also retrieves the same resource repeatedly, and does not re-fetch easily. The failure mode of fetch failing after retries can be an unrecoverable configuration error.

The API extension is to add a Blob resource reference and Blob config source to RemoteDataSource.

Blob discovery service client would be similar to RDS behavior. It would maintain a list of blobs indexed by references from Wasm and other extensions.

The behavior change in Wasm extensions would be:
- Wasm issues a subscription for new blob references and puts the underlying listener into warming at first
- On the first update, the listener is activated
- On subsequent updates, Wasm performs an internal update (how it does merits more discussion, since creation and draining of VMs is expensive)

It is worth mentioning that BDS can also subsume SDS since the payload is mostly opaque in SDS.

**Comment by mandarjog** [comment]
+ ADS can order BDS ahead of LDS that use them avoiding warming all together.

@jplevyak @PiotrSikora @htuch

**Comment by kyessenov** [comment]
@mandarjog Yeah that's true. We can treat warming similar to LDS/CDS interaction where listener awaits until the cluster is pushed in the state-of-the-world. I was initially thinking of LDS/RDS interaction where listener requests a particular route instead.

**Comment by htuch** [comment]
I think a blob discovery service is a great idea! I've long thought SDS should look more like this. Let's do it.

Re: warming, the problem with pushing BDS ahead is the proxy needs to be ready to receive it. In general, unless this is a SotW API, we expect that clients are free to ignore updates for resources that they don't have a reference to in their DAG, CC @markdroth

**Comment by PiotrSikora** [comment]
I thought that the main difference between BDS and SDS was that BDS was supposed to push resources ahead of time via ADS, whereas SDS requires proxy to explicitly request and subscribe to them.

If we're going to subscribe to resources and warm listeners, then this is literally SDS (unless I've missed something), so why not just us that?

**Comment by htuch** [comment]
If you want to do that, you are going to have to make BDS a SotW API which is subscribed to with wildcard (i.e. it's similar to LDS or CDS). Otherwise the xDS transport does not guarantee that a client will retain the resources.

**Comment by htuch** [comment]
Also, IMHO SDS is way too complicated and should be deprecated in favor of a simpler mechanism like BDS. It conflates transport level delivery of resources to the proxy with a bunch of rigid interfaces around certificate management.

**Comment by PiotrSikora** [comment]
@htuch I don't care too much one way or the other, but I'm confused how BDS with subscription is different and/or simpler than SDS.

**Comment by htuch** [comment]
SDS is packing together a bunch of resources, e.g. certificate chain, private key, and having to unpack them and then coordinate watches across on the filesystem, e.g. https://github.com/envoyproxy/envoy/blob/master/source/common/secret/sds_api.cc#L63 and https://github.com/envoyproxy/envoy/blob/master/source/common/secret/sds_api.cc#L133. 

I think we should have just a general purpose way of pushing blobs, and ideally use xDS filesystem transport rather than dedicated watches in SDS to provide secrets.

**Comment by mattklein123** [comment]
Overall I'm supportive, but I think there is some intentional overlap with https://github.com/envoyproxy/envoy/issues/7867, as I assume you implicitly want to update WASM code out of band from everything else (and possibly share it between multiple filter chains).

Can we please do a design doc to make sure we think through all of the various cases including potentially using this for SDS?

**Comment by htuch** [comment]
+1 on design doc. The way I think of this is that we're providing essentially a way for the management server to tell a proxy what blobs it might be interested in, essentially populating a virtual filesystem on the proxy. Then we can have any API, via a DataSource, refer to these blobs. 

This is very simple for SotW, but I think considering what would be needed for on-demand is worth doing. Because, there is an interplay between a wildcard SotW API with on-demand and the existing non-wildcard delta xDS. FWIW, this is all just because we have a confused notions of collections in the current xDS APIs; https://github.com/envoyproxy/envoy/issues/11264 will make this quite a bit saner.

**Comment by kyessenov** [comment]
Thanks for the feed. Yes, I am aware of the connection to filter config discovery service. I can imagine having FCDS working for Wasm would also solve the motivating problem. I am not entirely sure which way is better. Both have general advantages.

**Comment by mattklein123** [comment]
>  I can imagine having FCDS working for Wasm would also solve the motivating problem. I am not entirely sure which way is better. Both have general advantages.

My only main comment on this point would be that I know we need FCDS for a ton of different use cases. So if FCDS solves the problem it might be a better use of limited engineering resources.

**Comment by kyessenov** [comment]
Unless there's an objection, I'm in favor of pursuing filter config discovery service instead. Wasm requires a VM per config instance, so that puts a reasonable upper bound on the number of different configs (roughly, the bound is ~1K VMs which is an upper bound of possible configs to be shipped). Ultimately, it's more natural to map user input (config) to a unit of edit in Envoy xDS tree.

**Comment by mattklein123** [comment]
+1 to pursuing FCDS. We need this for quite a few different use cases at Lyft moving forward and can help out. cc @wgallagher @gkleiman @tonya11en @LisaLudique

**Comment by mandarjog** [comment]
FCDS is solving the general motivating problem, however it is not able to share the most expensive part of the config directly. For example if we have same config and Wasm bits but 50 different vmids, proxy will request essentially the same thing 50 times. 
FCDS is yet to be designed so we can ensure that this part it optimized. 

As an example: in a small istio deployment with 50 proxies,  the full push data spike is about 90kb. In case where we have just 5 filters with different vm ids, and size of a Wasm module is 1MB, we will increase the push from 90Kb to 5000Kb. 

FCDS can work in conjunction with Wasm cache, which would still be a win if we do not need to drain the listener.

**Comment by kyessenov** [comment]
@mandarjog Could we make it that Wasm filter reads VM_ID from other places, too? I think the number of 50 comes from tenants in your cases, and they are probably segregated by listeners. We could make the Wasm filter read a listener property to determine which VM_ID to use.

**Comment by mattklein123** [comment]
> FCDS is yet to be designed so we can ensure that this part it optimized.

I think it's possible that we could build a config sharing (reference) system into FCDS, but yes this needs to be thought through. I'm not sure what the "WASM cache" is and how it relates to this.

**Comment by mandarjog** [comment]
Wasm cache is something that we added to envoy-wasm while this solution is being worked out. During creation of a vm an in memory cache is checked to see if the bit are already downloaded. If the sha matches then we don’t download it again. This means that as long as the same wasm bits/sha is used it will not be carted around repeatedly.


https://github.com/envoyproxy/envoy-wasm/blob/cc7b50f6fe65a42534cf8f0e6e8ff784fbfe104c/source/extensions/common/wasm/wasm.cc#L228
